enhancements/monitoring/restart-metrics.md

```diff
@@ -0,0 +1,122 @@
+---
+title: restart-metrics
+authors:
+  - "@rphillips"
+reviewers:
+  - "@smarterclayton"
+  - "@sjenning"
+  - "@mrunalp"
+approvers:
+  - "@brancz"
+  - "@bparees"
+  - "@squat"
+  - "@s-urbaniak"
+  - "@metalmatze"
+  - "@paulfantom"
+  - "@LiliC"
+  - "@pgier"
+  - "@simonpasquier"
+creation-date: 2020-03-23
+last-updated: 2020-03-23
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Service Restart Metrics
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+There is not any insight with systemd units restarting. Metrics of this sort are
+important to know if Kubelet, crio, or other system services are crashing and
+restarting. Restart metrics are going to be extremely important since the
+Kubelet is going to be changed to exit on a crash, thus restarting. Previous
+behavior of the Kubelet is to recover() from the crash and not exit.
+
+## Motivation
+
+Systemd unit restart metrics of a unit are vitally important for system
+administrators to see the health of the Kubelet, crio, and other vitally important
+system services.
+
+### Goals
+
+- Systemd unit restart metrics need to be propogated through to monitoring, and
+  the alerting system to alert system administrators to problems within the
+  system.
+
+### Non-Goals
+
+- Defining alerts is not in the scope of this proposal. Alerts can be defined
+  once we see some sample cluster data.
+
+### Current Behavior
+
+node_exporter is running in a non-privileged container - as user `nobody` - and
+cannot connect to the DBUS socket to communicate with systemd. Monitoring has
+concerns the underlying [systemd
+collector](https://github.com/prometheus/node_exporter/blob/master/collector/systemd_linux.go)
+is not performant enough.
+
+## Proposal
+
+- This proposal is to add a privileged sidecar to node_exporter running
+  [systemd_exporter](https://github.com/povilasv/systemd_exporter])
+
+- systemd_exporter will be configured to write metrics to node_exporters
+  text_collector.
+
+- A whitelist will be enabled in systemd_exporter to enable metrics from the
+  following core services:
```

Comment:
I feel like _at least_ the services listed here deserve to be monitored in a whitebox fashion, not just indirectly through the proposed mechanism. Can we ensure that as a follow up of this "generic" monitoring, we also add whitebox monitoring to the node team and/or coreos team's backlogs? kubelet and crio are already, and chrony would have been helpful multiple times already and I imagine the same holds true for all other services listed here (I believe I've talked about NetworkManager with @lucab before as well).

Reply:
@brancz I can add a section on what we would like to add later... Can you provide me some examples of the whitebox testing you would like to see?

Reply:
Essentially each of these components should either have a metrics endpoint of their own or we should write an exporter for it. kubelet and crio already do so that's good, but all the others in this list should as well or at least have an exporter.

Reply:
Sorry I can't seem to find where this is documented. Can you point me at it?

Reply:
Most of these other components are system components, and the restart metric is likely a good indicator of failure.

Reply:
@brancz 
Do you have more question ?

Thanks